src/activity_handlers.cpp

```diff
@@ -2781,3 +2783,77 @@ void activity_handlers::haircut_finish( player_activity *act, player *p ) {
     p->add_morale( MORALE_HAIRCUT, 3, 3, 480_minutes, 3_minutes );
     act->set_to_null();
 }
+
+static std::vector<tripoint> get_sorted_tiles_by_distance( const tripoint abspos,
+        const std::unordered_set<tripoint> &tiles )
+{
+    auto cmp = [abspos]( tripoint a, tripoint b ) {
+        int da = rl_dist( abspos, a );
+        int db = rl_dist( abspos, b );
+
+        return da < db;
+    };
+
+    std::set<tripoint, decltype( cmp )> sorted( tiles.begin(), tiles.end(), cmp );
+    std::vector<tripoint> vector( sorted.begin(), sorted.end() );
+
+    return vector;
+}
+
+template<typename fn>
+static void cleanup_tiles( std::unordered_set<tripoint> &tiles, fn &cleanup )
+{
+    auto it = tiles.begin();
+    while( it != tiles.end() ) {
+        auto current = it++;
+
+        const auto &tile_loc = g->m.getlocal( *current );
+
+        if( cleanup( tile_loc ) ) {
+            tiles.erase( current );
+        }
+    }
+}
+
+void activity_handlers::till_plot_do_turn( player_activity*, player *p )
+{
+    const auto &mgr = zone_manager::get_manager();
+    const auto abspos = g->m.getabs( p->pos() );
+    auto unsorted_tiles = mgr.get_near( zone_type_id( "FARM_PLOT" ), abspos );
+
+    // Nuke the current activity, leaving the backlog alone.
+    p->activity = player_activity();
+
+    // cleanup unwanted tiles
+    auto cleanup = [p]( const tripoint & tile ) {
+        return !p->sees( tile ) || !g->m.has_flag( "DIGGABLE", tile ) || g->m.has_flag( "PLANT", tile ) ||
+               g->m.ter( tile ) == t_dirtmound;
+    };
+    cleanup_tiles( unsorted_tiles, cleanup );
+
+    // sort remaining tiles by distance
+    const auto &tiles = get_sorted_tiles_by_distance( abspos, unsorted_tiles );
+
+    for( auto &tile : tiles ) {
+        const auto &tile_loc = g->m.getlocal( tile );
+
+        auto route = g->m.route( p->pos(), tile_loc, p->get_pathfinding_settings(), p->get_path_avoid() );
+        if( route.size() > 1 ) {
+            route.pop_back();
+            p->set_destination( route, player_activity( activity_id( "ACT_TILL_PLOT" ) ) );
+            return;
+        } else { // we are at destination already
+            p->add_msg_if_player( _( "You churn up the earth here." ) );
+            p->moves = -300;
```

Comment:
I believe this line should be:

`p->moves -= 300;` - i.e. subtract 300 from player moves.

Now it sets player moves to -300 which makes following check (`if( p->moves <= 0 ) {`) redundant as this expression will always be true.

Reply:
Yeah, I think you are correct, that was my intended behavior... not that I came with the formula, I stole these 3 lines from iuse::makemound and I did not check them properly I guess. I think it should be changed in both places. I would normally not duplicate the code, but I did, as this was only 3 lines.

Should I make a PR to add a cost to game_constants.h and use the value in both places? I think it is better than making extra function with just 3 lines of code.

Reply:
No new function is necessary right now and you can simply change values in existing places.

Btw, you can also use following syntax: `p->mod_moves( -300 );`

Reply:
Fix is in #25253 , tested both manual and automatic tilling.